lib/llvm/EmitGCLayoutInfo.cpp

```diff
@@ -0,0 +1,133 @@
+#include "kllvm/ast/AST.h"
+#include "llvm/IR/Instructions.h"
+#include "llvm/IR/LegacyPassManager.h"
+#include "llvm/IR/Module.h"
+#include "llvm/IR/Statepoint.h"
+#include "llvm/Support/raw_ostream.h"
+#include <cstdlib>
+#include <vector>
+
+using namespace llvm;
+using namespace kllvm;
+
+#define DEBUG_TYPE "EmitGCLayoutInfo"
+
+namespace {
+struct EmitGCLayoutInfo : public ModulePass {
+  static char ID;
+  EmitGCLayoutInfo()
+      : ModulePass(ID) { }
+
+  void error(Type *Ty) {
+    llvm::errs() << "Could not identify garbage collection information! This "
+                    "is a bug in the llvm backend of K\n";
+    Ty->print(llvm::errs());
+    abort();
+  }
+
+  bool runOnModule(Module &M) override {
+    unsigned int nextID = 0;
+    unsigned int numRelocations = 0;
+    std::map<std::pair<unsigned int, unsigned int>, ValueType> cats;
+    auto &Ctx = M.getContext();
+    for (Function &F : M.functions()) {
+      for (BasicBlock &BB : F.getBasicBlockList()) {
+        for (Instruction &I : BB.getInstList()) {
```

Comment:
A very minor issue, but you can iterate directly over Modules / Functions / BasicBlocks rather than calling these functions.

```suggestion
    for (Function &F : M) {
      for (BasicBlock &BB : F) {
        for (Instruction &I : BB) {
```